031-statefulset-removal.md

```diff
@@ -0,0 +1,231 @@
+Users will need to understand the consequences of the operator not running and managing the pods.
+
+Unlike StatefulSets, KafkaNodeSet also doesn't use template to generate Pods.
+That brings us many advantages such as asymmetric configuration etc.
+But it also means that the custom resource will be bigger than StatefulSets because it contains a full YAML for each Pod.
+It will have a limited capacity how many Pods it will be able to handle.
+But since the KafkaNodeSets do not enforce any specific pod names etc., we should be able to workaround this by having more KafkaNodeSets if needed.
+
+Performance testing would need to be done once a full implementation exists to ensure that the KafkaNodeSet controller is able to handle big Kubernetes clusters with many pods belonging to different Kafka clusters as well as unrelated pods.
+
+## Affected components
+
+This proposal currently discuss only the removal of StatefulSets used for Kafka and ZooKeeper.
+While in theory it could also be used to replace Deployments used for other components, this is not covered by this proposal.
+
+## Rejected alternatives
+
+### _Single node_ StatefulSets
+
+One of the things considered was to have a StatefulSet per pod.
+The StatefulSet would still have the same limitations as before.
+But with many different StatefulSets and each having only one Pod, we would not need to care.
+However, this approach makes the implementation more complex.
+For example:
+* Every pod would be named something like `my-cluster-kafka-1001-0` where `1001` would be an example of a broker ID and `0` is the suffix given to the Pod by the StatefulSet
+* We would need to manage multiple resources
+
+The main advantage of this approach is that Kubernetes would be still responsible for creating the Pods and we would not need to worry about it.
+
+### Managing the pods directly
+
+It would be also possible to manage the Pods from the `KafkaAssemblyOperator`.
+In the same way as it is currently creating the StatefulSets or deleting the pods, it can also create and update the Pods directly.
+This would make things easier because we would not need any additional controller or custom resource.
+However, the Kafka CR reconciliation is currently very _heavy_ process which does lot of things in a serial fashion.
+It manages both ZooKeeper and Kafka and it uses long-running reconciliation.
+The reconciliation often takes minutes while it waits for the Pods getting ready or for load balancers being created.
+It runs multiple reconciliations in parallel, but for each Kafka custom resource only one reconciliation can be in progress.
+That means that it might often take a lot of time before the reconciliation gets to the point when it can recreate the Pods.
+Imagine situation when several Kafka brokers are gone because of failure in one of the data centers but the reconciliation is stuck waiting on some load balancer creation to finish while the pods are missing.
```

Comment:
```suggestion
Imagine a situation in which several Kafka brokers are unavailable because of a failure in one of the data centers, and the reconciliation is stuck waiting for a load balancer creation to finish while the pods are missing.
```